File: lib-segment/src/main/java/websiteschema/mpsegment/hmm/Pi.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package websiteschema.mpsegment.hmm;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import websiteschema.mpsegment.util.ISerialize;
import websiteschema.mpsegment.util.SerializeHandler;

/**
 *
 * @author ray
 */
public class Pi implements ISerialize {

    private Map<Integer, Double> pi = new HashMap<Integer, Double>();
    private int total = 1;

    public Pi() {
    }

    public Pi(Map<Integer, Integer> pii) {
        init(pii);
    }
    
    private void init(Map<Integer, Integer> pii) {
        Set<Integer> keySet = pii.keySet();
        for (Integer i : keySet) {
            total += pii.get(i);
        }

        for (Integer i : keySet) {
            setPi(i, (double) pii.get(i) / (double) (total));
        }
    }

    public double getPi(int index) {
        if (pi.containsKey(index)) {
            return pi.get(index);
        } else {
            return 1.0 / (double) total;
        }
    }

    public void setPi(int index, double prob) {
        pi.put(index, prob);
    }

    @Override
    public void save(SerializeHandler writeHandler) throws IOException {
        writeHandler.serializeInt(total);
        writeHandler.serializeMapIntDouble(pi);
    }

    @Override
    public void load(SerializeHandler readHandler) throws IOException {
        total = readHandler.deserializeInt();
        pi = readHandler.deserializeMapIntDouble();
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/hmm/Pi.scala
package websiteschema.mpsegment.hmm

import websiteschema.mpsegment.util.ISerialize
import websiteschema.mpsegment.util.SerializeHandler
import collection.mutable.HashMap
import collection.mutable.Map

class Pi extends ISerialize {

    private var pi: Map[Int, Double] = HashMap[Int,Double]()
    private var total = 1

    private def init(pii: Map[Int,Int]) {
        val keySet = pii.keys
        for (i <- keySet) {
            total += pii(i)
        }

        for (i <- keySet) {
            setPi(i, pii(i).toDouble / total.toDouble)
        }
    }

    def getPi(index: Int) : Double =
      pi.get(index) match {
        case Some(p) => p
        case _ => 1.0D / total.toDouble
      }

    def setPi(index: Int, prob: Double) {
        pi += (index -> prob)
    }

    override def save(writeHandler: SerializeHandler) {
        writeHandler.serializeInt(total)
        writeHandler.serializeMapIntDouble(pi)
    }

    override def load(readHandler: SerializeHandler) {
        total = readHandler.deserializeInt()
        pi = readHandler.deserializeMapIntDouble()
    }
}

object Pi {

  def apply() = new Pi()

  def apply(pii: Map[Int, Int]) = {
    val pi = new Pi()
    pi.init(pii)
    pi
  }

}
